---
title: "DESeq2 Tutorial"
author: "Charlie George & Modified by Devika Agarwal"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources 

- **DESeq2 Vignette** - http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- **Bioconductor RNAseq Workflow Book** - http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html


# install packages 

- If i didn't already have these installed in my environment - I would delete the # on lines 19-23 and 26-39 and run the code 

```{r}
#from cran 
install.packages('tidyverse')
install.packages("RColorBrewer")
install.packages('pheatmap')
if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#from bioconductor
BiocManager::install("ashr")
BiocManager::install("apeglm")
BiocManager::install("DESeq2")
BiocManager::install("org.Mm.eg.db")
```

# load packages

- what do these packages do? Add comments to remind you what they do 
```{r}
library("pheatmap")   
library("RColorBrewer")
library(tidyverse)     
library(DESeq2)      
library(ashr)          
library(apeglm)      
library(org.Mm.eg.db)
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

- We made this (or a very similar version) using featurecounts in week1 on the cluster
- raw data -> fastqc -> hisat2 -> featurecounts 

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")
head(featurecounts_table)

# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)
head(gene_info)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  dplyr::select(!c(Chr, Start, End, Strand, Length)) %>%
                  as.matrix()

# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 
dim(counts_table)
head(counts_table)

```

# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info.tsv")
sample_info

# What class is the cell_type currently? 
class(sample_info$cell_type) #it is character string type


# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 
# levels are categories, in this dataset our levels are our cell_type, where we have CD4+ cells as a control level, and CD8+ cells as a test

sample_info$cell_type <-factor(sample_info$cell_type, levels = c("CD4", "CD8")) 
head(sample_info)

# what other columns might we want to include in our design that we want to change to factors? 
# replicate and genotype

# use as.factor here to let it set the levels seperately, so it can automatically asign it based on the information in the column, where it pulls out data in column
sample_info$genotype <-as.factor(sample_info$genotype) 

# factor you can specify your own levels - more explicite
sample_info$replicate <-factor(sample_info$replicate, levels = c("1", "2", "3"))

head(sample_info)


```

# Check the sample names are the same and in the same order in the sample_info and counts_table 

```{r}
# Column names of the count table and row names of the sample info table must be in the same order
# Check this 

colnames(counts_table)
rownames(sample_info)

# this returns lots of falses, suggesting the colnames do not equal the rownames
colnames(counts_table) == rownames(sample_info)

# make sample_info row names from sample_id 

head(sample_info)
rownames(sample_info) <- (sample_info$sample_id) # this doesn't work because it's a tibble

# first we need to convert sample_info from a tibble to a dataframe, as tibbles can't have rownames
sample_info <- sample_info %>% 
  as.data.frame() %>% 
  column_to_rownames("sample_id")

head(sample_info)

# remove .bam from colnames of the counts_table - use str_replace function with the pattern you want to remove, followed by what you want to replace it with, use "" if nothing.
colnames(counts_table) 

colnames(counts_table) <- str_replace(colnames(counts_table), pattern = ".bam", "")
colnames(counts_table) 


# check again whether your column names in the counts table is equal to the row names in the sample info
colnames(counts_table) == rownames(sample_info) 

```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
#DESeqDataSetFromMatrix(countData,colData,design,tidy = FALSE,ignoreRank = FALSE,...)
# design here, means were controlling for the replicate and the genotype and then finally looking for differences between the cell type

dds <- DESeqDataSetFromMatrix(countData = counts_table, colData = sample_info, design = ~replicate + genotype + cell_type)
  
# creating another dds object that stores the gene_info in row data too
dds2 <- DESeqDataSetFromMatrix(countData = counts_table, colData = sample_info, rowData = gene_info, design = ~replicate + genotype + cell_type)


```

## Investigate the dds object 

```{r}
# what does it look like? - e.g. use str(dds)

str(dds)
str(dds2)

# how to access counts

counts(dds)
assays(dds)$counts # alternative
assay(dds, "counts") # other alternative

# access colData i.e the sample info table

colData(dds)

# access rowData

rowData(dds) # this slot is empty for this as we didn't assign anything
rowData(dds2) # row data is now present

# access the design

design(dds)

# does it have any rowRanges()?

rowRanges(dds)
rowRanges(dds2)

# what are its rownames and colnames?

rownames(dds) # ensembl gene identifier from counts_table
colnames(dds) # sample id from counts_table

# what attributes does it have?

attributes(dds) # dispersion function is currently empty- but we'll come onto that next

```

## How to subset a DESeqDataSet Object 

```{r}
# practice subsetting dds object
# select first 3 columns

dds[,1:3] # this gives you all rows and 3 columns

```

# Filter out unexpressed genes 

- It's good practice to filter out genes with 0 counts across all samples at the start 
- This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis 
- Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.


```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`

dds_filtered_true_false <- rowSums(counts(dds)) > 0
dds_filtered_true_false

dds_filtered <- dds[rowSums(counts(dds)) > 0, ]
dds_filtered

# check dims of dds_filtered  

dim(dds_filtered)
  
```

# Perform exploratory analysis on dds_filtered object 

1. transform dds_filtered object using `rlog()` functions
    - name your new variables `rlog_transformed_dds` 
    - we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    - check the class of the rlog_transformed_dds objects 
    - you can look at the tranformed counts table using `assay()` 

NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!


```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind=TRUE)
rlog_transformed_dds


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 
class(rlog_transformed_dds)
  
#get a look at the transformed counts table using `assay()`
assay(rlog_transformed_dds)

```

# Optional: If you have time do the tranformation again but using the vst() function 

- As a guide 
    - rlog - more robust than VST when size factors vary widely (10x)
    - VST - less sensitive to high count outliers than rlog, quicker - recommended most often 

```{r}
# get vst transformed data using vst() - alternative method function for transforming data

vst_transformed_data <- vst(dds_filtered, blind = TRUE)
vst_transformed_data

class(vst_transformed_data)

```


# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()


```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 

corr_dds_spearman <- cor(assay(rlog_transformed_dds), method="spearman")

corr_dds_pearson <- cor(assay(rlog_transformed_dds), method="pearson")

corr_dds_spearman
corr_dds_pearson


# use pheatmap to plot the correlation matrix 
# the issue with this is that the scale is very small 0.998 to 1, so instead we can use a distance matrix

pheatmap(corr_dds_spearman)

pheatmap(corr_dds_pearson)

```

# try ploting a heatmap of sample distances 

```{r}
# get sample distances -> Annotated what this code is doing?
sampleDists <- dist(t(assay(rlog_transformed_dds))) 
sample_distances_matrix <- as.matrix(sampleDists) # turning it into a matrix so we can plot it

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255) # just changing the colours

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)

# what does the 0 to 1 scale mean? Need to remove samples two suspicious samples ending in 88 and 89
```

What do you notice here!? what does it tell us about our data? 

Answer: Blocks of 4 dark blue - this isn't right, as they're coming up with zero counts each time. The suspicious samples are associated with 089 and 088, so we need to remove these. 



## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
colnames(rlog_transformed_dds)

# see that we can just take the first six columns to remove the suspicious samples
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6] 

# alternative - this is much more complicated, but would need to use if we couldn't easy subset out
rlog_transformed_dds_subset3 <- as.data.frame(assay(rlog_transformed_dds)) %>% dplyr::select(-"ERR1755088", -"ERR1755089") 
rlog_transformed_dds_subset3

# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above
sampleDists <- dist(t(assay(rlog_transformed_dds_subset))) 
sample_distances_matrix <- as.matrix(sampleDists) # turning it into a matrix so we can plot it

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255) # just changing the colours

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)

```

What do you notice now about the data and the clustering?


# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r, fig.asp=1}

# plot pca of rlog_transformed_dds using plotPCA()
# plotPCA(object, intgroup = "condition", ntop = 500, returnData = FALSE)
# ntop = 500 is default
# it's very small because the scales of PC1 (99%) and PC2 (0%) are very far apart, this isn't normally the case. R is trying to be clever.


plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type", ntop = 500) 


# what does the intgroup parameter do? have a go at changing it...
# if i change it to genotype the pattern doesn't change but the colour changes because they are all KO - and you didn't assign this as a comparison earlier in the design

plotPCA(rlog_transformed_dds_subset, intgroup = "genotype", ntop = 500)

# how many variable genes is it using for pca? 
# is it using 500 as ntop is 500: from help page "ntop: number of top genes to use for principal components, selected by highest row variance"

plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type", ntop = 100)

plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type", ntop = 1000)

plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type", ntop = 500)

# plot the pca for all genes 

plotPCA(rlog_transformed_dds_subset, intgroup = "cell_type", ntop = 1000)

```


What does it tell us about our sample?

Answer: 


------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis 
    e.g. - update design with variables in sample info table we want to regress out 
         - remove samples (poor quaility or those with name switches etc)
         
# Revisiting our inital data 

What do we need to do to our data??? 

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 
# we're selecting the first 6 columns 
dds_filtered
dds_filtered_updated <- dds_filtered[,1:6] 
dds_filtered_updated # moves from 8 columns to 6

# do we want to update the design? Is this necessary in this case?
# update the design to control for replicate and examine the effects of cell_type
# e.g. design(dds) <- ~replicate + condition

design(dds_filtered_updated) <- ~replicate + cell_type

```


# Now we can start our actual DE!!!! yay 

Run deseq2 modelling using one function (DESeq())

```{r}

# look at dds slots then run deseq2 function 
# what has happened to your dds slots?
# run the dds alone, then run the DESeq() and see what has happened to the slots
# object has lots of different slots that we populate with information

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated)

# sumarise results 
summary(results_obj)

```

Have a look at the outputs 

# Now we will run the analysis step by step to see whats going on 

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()

dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)

# Access the size factors from the dds_filtered_updated object

sizeFactors(dds_filtered_updated)

# Do you see any systematic differences in size factors between groups? First 3 tend to be larger and the last 3 tend to be smaller.

```

# Optional: Plot size factors as barchart using ggplot
```{r}

size_factors <- sizeFactors(dds_filtered_updated)
size_factors <- data.frame(sample = names(size_factors),
                           size_factor = size_factors,
                           sample_group = colData(dds_filtered_updated)$cell_type)

ggplot(size_factors, aes(x = sample, y = size_factor, fill = sample_group)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 45, h = 1))


```


## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts

normalised_counts <- counts(dds_filtered_updated, normalized=TRUE)
normalised_counts

# check the class of the output

class(normalised_counts)

```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions()
# have to reassign again to update the object with those dispersions
# note dds_filtered is excluding those samples with counts =0, then dds_filtered_updated is removing the dodgy samples

dds_filtered_updated <- estimateDispersions(dds_filtered_updated)
dds_filtered_updated

# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
# this is the plot where you can see how the data points look: watch out for if the black dots are far apart, if the line dips in the middle then goes up again - these are signs the data aren't great. This dispersion plot looks good. 

plotDispEsts(dds_filtered_updated)


```


## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest
# by assigning it back to dds_filtered_updated, it keeps updating it, it's not overwriting it, just adding more information. 
# then you create a new results object that you can do other specific things to (see below)

dds_filtered_updated <- nbinomWaldTest(dds_filtered_updated)

res <- results(dds_filtered_updated)

```


# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# these are the main important steps we've done: this is a summary of everything you've done, so you can see it all in one place

# generate dds object 

dds_actual_analysis <- DESeqDataSetFromMatrix(countData = counts_table, colData = sample_info, design = ~replicate + cell_type)

# filter rows with 0 counts 

dds_actual_analysis <- dds_actual_analysis[rowSums(counts(dds_actual_analysis)) > 0, ]
dds_actual_analysis

# remove unwanted samples (which we identified with heatmap)

dds_actual_analysis <- dds_actual_analysis[,1:6]  # moves from 8 columns to 6

# run DESeq() 

dds_actual_analysis <- DESeq(dds_actual_analysis)
str(dds_actual_analysis)

```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify?
# can change arguments in the results() function. Change alpha to 0.05 (default is 0.1)
# default is usually fine, e.g. pAdjustMethod is Benjamini Hochberg - don't need to change this
# contrast may be the other thing you want to change - conditions to compare: contrast = c("cell_type", "CD4", "CD8") - the first thing is the thing you're comparing (the thing that you're interested in - in this case the cell_type and CD4) then the last is the reference (in this case CD8 is what we're using as the baseline). So for my results this would be maybe contrast = c("genotype", "M337V", "WT")

res_object <- results(dds_actual_analysis, contrast = c("cell_type", "CD8", "CD4"), alpha = 0.01)
summary(res_object)
res_object

```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vingette("DESeq2")
```{r}

# unrelated but if you want to save anything in an object you need to pull it out as a table first then do write.tsv tidyverse, or can save it directly if it's a table. Usually you only save the output you generate at the end. 

coef_dds <- coef(dds_actual_analysis)
head(coef(dds_actual_analysis))

```

# Plot normalised counts for ENSMUSG00000041779
 
 checkout the 'plotCounts()` function
 
```{r}
# Plot normalised counts for gene ENSMUSG00000041779

# first recalculate normalised counts using dds_actual_analysis

normalised_counts <- counts(dds_actual_analysis, normalized=TRUE)

# Then pull out a specific gene

normalised_counts["ENSMUSG00000041779", ]

# Then generate a gene count dataframe of the normalised counts from the gene of interest and change the rownames to a column called sample

gene_count <- normalised_counts["ENSMUSG00000041779", ] %>% 
  as.data.frame() %>% 
  rownames_to_column("sample")
gene_count

# rename the column of counts from . to count

colnames(gene_count) <- c("sample", "count")
gene_count

# then we need to add CD8 and CD4 labels to our table, as we currently have nothing to group it by. The first 3 are CD4 and the second 3 are CD8

colData(dds_actual_analysis) # this is where we can find the column with cell_types that we need

# Then pull out this column and assign the name of the cell_type column to cell_type using the mutate function

gene_count <- gene_count %>% 
  mutate("cell_type" = colData(dds_actual_analysis)$cell_type)
gene_count


# then plot the gene count

gene_count %>% 
  ggplot(aes(x=cell_type, y=count)) +
  geom_boxplot() +
  geom_point()


```

# Access the results table for the comparison between CD8 and CD4 T cells . You can also try to do CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control 

checkout `resultsNames()`  & `results()`

```{r}

# access the results table for the comparison between CD8 and CD4 T cells (can do CD4 vs CD8, change the order to switch CD8 to reference)

res_object <- results(dds_actual_analysis, contrast = c("cell_type", "CD8", "CD4"), alpha = 0.01)
res_object

# do the same thing but change the replicate, then change replicate 3 to the control vs replicate 1

res_object_replicate <- results(dds_actual_analysis, contrast = c("replicate", "1", "3"), alpha = 0.01)
head(res_object_replicate)

```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected? for CD8 vs CD4

```{r}

res_object %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_histogram(aes(x=pvalue), fill='lightblue') + 
    geom_histogram(aes(x=padj),fill='lightgreen')

head(res_object)

```


# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
# MA plot: each blue dot shows the genes which are significantly up regulated and which are significantly downregulated, in grey are the genes which are not significantly different. This is showing you the differences in gene expression between cell_type CD8 and CD4
# this function requires an object, rather than a dataframe

plotMA(res_object, alpha = 0.01, ylim = c(-10,10))

```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunkenLFC_dds_results_obj

resultsNames(dds_actual_analysis) # use this to get the coefficient
shrunkenLFC_dds_results_obj <- lfcShrink(dds_actual_analysis, coef="cell_type_CD8_vs_CD4", type = "apeglm")

# then replot the MA plot with the shrunken data - just makes it easier to see by reducing the number of grey dots for genes on the left of the plot

plotMA(shrunkenLFC_dds_results_obj,alpha = 0.01, ylim = c(-10,10)) 

# other method?
lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="ashr") %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5, ColNonSig = "blue", ColSig = "red") 

```


# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids 

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df

shrunkenLFC_dds_results_df <- as.data.frame(shrunkenLFC_dds_results_obj)

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 
# use keytypes command to remind us of what keytypes are in this particular db - tells us what we can add to our table, the most useful things are Ensembl IDs (most unique ID - each gene is different), unlike gene symbol where you might have multiple genes called the same thing

keytypes(org.Mm.eg.db) 

# How many Ensembl IDs are not assigned a gene symbol and how many gene symbols are duplicated?
# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI
# add mapId vectors to  shrunkenLFC_dds_results_df
# # mapIds gives you a 1:1 mapping by default (can only extract one column at a time), whereas the select function doesn't (but you can extract multiple columns at a time)

MGI_symbols <- mapIds(org.Mm.eg.db, row.names(shrunkenLFC_dds_results_df), "MGI", "ENSEMBL")
gene_symbols <- mapIds(org.Mm.eg.db, row.names(shrunkenLFC_dds_results_df), "SYMBOL", "ENSEMBL")
entrez_ids <- mapIds(org.Mm.eg.db, row.names(shrunkenLFC_dds_results_df), "ENTREZID", "ENSEMBL")


#to check that the order of the vectors are the same as the df
table(names(entrez_ids)==row.names(shrunkenLFC_dds_results_df))    
#options(scipen=0)

#add mapId vectors to the original results df
results_shrunken_LFC_apeglm_filtered_df <- shrunkenLFC_dds_results_df %>%
    as.data.frame() %>% 
    mutate(mgi_symbols=MGI_symbols) %>%
    mutate(gene_symbols=gene_symbols) %>%
    mutate(ENTREZ_IDs=entrez_ids)


# How many Ensembl IDs are not assigned an mgi symbol and how many mgi symbols are duplicated?

which(duplicated(MGI_symbols))# output will show the duplicates


no_symbols <- which(is.na(MGI_symbols))# shows which genes don't have any symbols associated with them (these genes don't have an MGI symbol - lots won't have annotations)
length(no_symbols)

```


# Export all the results to CSV

```{r}
results_shrunken_LFC_apeglm_filtered_df <- results_shrunken_LFC_apeglm_filtered_df %>%  
  rownames_to_column("ensembl_gene_id")

write_csv(results_shrunken_LFC_apeglm_filtered_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")

# if you're in the right place this is fine, but good to specify the path if you're not sure
# write_csv(results_shrunken_LFC_apeglm_filtered_df, "/project/obds/ecarroll/deseq2_tutorial/CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv"


```

# Filter res_df and save as CSV

```{r}

results_shrunken_LFC_apeglm_filtered_df <- results_shrunken_LFC_apeglm_filtered_df %>%  
  rownames_to_column("ensembl_gene_id")

#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```


Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....

