{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "61bd2c8d-8e1b-4dcb-b6f3-b2f2c5cf0d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "522a8e97-a02e-4aee-80d8-5bf591e763e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/dk/dft1sb753rzdxzmxscz2nqf00000gn/T/ipykernel_1483/3810453286.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('Homo_sapiens.GRCh38.107.gtf.gz',\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0               1            2        3        4  5  6  7  \\\n",
       "0  1  ensembl_havana         gene  1471765  1497848  .  +  .   \n",
       "1  1  ensembl_havana   transcript  1471765  1497848  .  +  .   \n",
       "2  1  ensembl_havana         exon  1471765  1472089  .  +  .   \n",
       "3  1  ensembl_havana          CDS  1471885  1472089  .  +  0   \n",
       "4  1  ensembl_havana  start_codon  1471885  1471887  .  +  0   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('Homo_sapiens.GRCh38.107.gtf.gz',\n",
    "                 sep =\"\\t\", \n",
    "                 header = None, \n",
    "                 comment='#', \n",
    "                 compression='gzip')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99257f0d-11a2-448d-b46b-5a483e222072",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'group']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "37f5c80e-0cfa-4bb1-8306-0aeef8b4422e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371239</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>32373</td>\n",
       "      <td>32528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371240</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>31698</td>\n",
       "      <td>31841</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000277475\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371241</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371242</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371243</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000275405\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3371244 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname          source          feature    start      end score  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848     .   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848     .   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089     .   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089     .   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887     .   \n",
       "...             ...             ...              ...      ...      ...   ...   \n",
       "3371239  KI270713.1         ensembl   five_prime_utr    32373    32528     .   \n",
       "3371240  KI270713.1         ensembl  three_prime_utr    31698    31841     .   \n",
       "3371241  KI270713.1         ensembl             gene    21861    22024     .   \n",
       "3371242  KI270713.1         ensembl       transcript    21861    22024     .   \n",
       "3371243  KI270713.1         ensembl             exon    21861    22024     .   \n",
       "\n",
       "        strand frame                                              group  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3371239      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371240      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371241      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...  \n",
       "3371242      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "3371243      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "\n",
       "[3371244 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "179530ee-0380-4bea-9683-d03fa7447987",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "havana                   2340968\n",
       "ensembl_havana            714608\n",
       "havana_tagene             170046\n",
       "ensembl                   139551\n",
       "mirbase                     5631\n",
       "ensembl_havana_tagene        275\n",
       "insdc                        144\n",
       "ensembl_tagene                21\n",
       "Name: source, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"source\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "91cf1097-7f29-43be-9f47-f35d5897b4ef",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'df'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py:3621\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3620\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3621\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_libs/index.pyx:136\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_libs/index.pyx:163\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'df'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [9]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m ens_df \u001b[38;5;241m=\u001b[39m df[\u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mcontains(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mensembl_havana\u001b[39m\u001b[38;5;124m\"\u001b[39m)]\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/frame.py:3505\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3503\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3504\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3505\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3506\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3507\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py:3623\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3621\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3622\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3623\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3624\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3625\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3626\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3627\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3628\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'df'"
     ]
    }
   ],
   "source": [
    "ens_df = df[df[\"df\"].str.contains(\"ensembl_havana\")].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dd7bdc87-1c41-49f8-b507-9fdcf308c55a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_df = df.query(\"source == 'ensembl_havana'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "052a8ebc-786c-4ebb-b4ba-0dce4b7f3aaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 714608 entries, 0 to 3370048\n",
      "Data columns (total 9 columns):\n",
      " #   Column   Non-Null Count   Dtype \n",
      "---  ------   --------------   ----- \n",
      " 0   seqname  714608 non-null  object\n",
      " 1   source   714608 non-null  object\n",
      " 2   feature  714608 non-null  object\n",
      " 3   start    714608 non-null  int64 \n",
      " 4   end      714608 non-null  int64 \n",
      " 5   score    714608 non-null  object\n",
      " 6   strand   714608 non-null  object\n",
      " 7   frame    714608 non-null  object\n",
      " 8   group    714608 non-null  object\n",
      "dtypes: int64(2), object(7)\n",
      "memory usage: 54.5+ MB\n"
     ]
    }
   ],
   "source": [
    "ens_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76abbfb0-a76b-4bec-a3e0-6460e2d76290",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method DataFrame.info of             seqname          source          feature    start      end score  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848     .   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848     .   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089     .   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089     .   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887     .   \n",
       "...             ...             ...              ...      ...      ...   ...   \n",
       "3371239  KI270713.1         ensembl   five_prime_utr    32373    32528     .   \n",
       "3371240  KI270713.1         ensembl  three_prime_utr    31698    31841     .   \n",
       "3371241  KI270713.1         ensembl             gene    21861    22024     .   \n",
       "3371242  KI270713.1         ensembl       transcript    21861    22024     .   \n",
       "3371243  KI270713.1         ensembl             exon    21861    22024     .   \n",
       "\n",
       "        strand frame                                              group  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3371239      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371240      -     .  gene_id \"ENSG00000277475\"; gene_version \"1\"; t...  \n",
       "3371241      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; g...  \n",
       "3371242      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "3371243      -     .  gene_id \"ENSG00000275405\"; gene_version \"1\"; t...  \n",
       "\n",
       "[3371244 rows x 9 columns]>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "12619d32-43b8-4666-b746-c78cb8877ab9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_df_feature = ens_df.groupby(\"feature\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b1ddee09-79a5-49f8-94d1-860528af9155",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f90af9e49a0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df_feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "95141b4a-9130-4915-94c3-5b51e7d10d3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon               285809\n",
       "CDS                262659\n",
       "five_prime_utr      38184\n",
       "three_prime_utr     30118\n",
       "transcript          26745\n",
       "stop_codon          25906\n",
       "start_codon         25889\n",
       "gene                19285\n",
       "Selenocysteine         13\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df[\"feature\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b665a938-4f71-4f21-8955-82637c277018",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ens_df.query(\"feature == 'gene'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ebb95fd8-e638-4eb2-9dd3-d242dd89efdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19285 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                              group  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...  \n",
       "...       ...                                                ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "\n",
       "[19285 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "28da574d-578c-4b52-9cc1-cf34258246dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/dk/dft1sb753rzdxzmxscz2nqf00000gn/T/ipykernel_1483/2396139549.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes['gene_length'] = df['end'] - df['start']\n"
     ]
    }
   ],
   "source": [
    "genes['gene_length'] = df['end'] - df['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "65279adc-eea4-4298-b856-c5441ed33166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19285 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                              group  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470  \n",
       "...       ...                                                ...          ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645  \n",
       "\n",
       "[19285 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "dddb5776-16f3-4a73-8538-e5b64d881699",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1.928500e+04\n",
       "mean     6.767102e+04\n",
       "std      1.298531e+05\n",
       "min      6.200000e+01\n",
       "25%      9.736000e+03\n",
       "50%      2.736400e+04\n",
       "75%      7.086500e+04\n",
       "max      2.473538e+06\n",
       "Name: gene_length, dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[\"gene_length\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b11d0602-c362-467f-be42-687ae4714f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_genes = genes.sort_values(by=['gene_length'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "901a71db-fe2a-4af0-a448-8cefee5abb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "longest_genes = sorted_genes.iloc[0:10,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "db2ad28f-b047-414e-a4cc-08bbd82b402d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601567</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2935353</td>\n",
       "      <td>4994972</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183117\"; gene_version \"20\"; ...</td>\n",
       "      <td>2059619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030998</th>\n",
       "      <td>20</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>13995369</td>\n",
       "      <td>16053197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000172264\"; gene_version \"18\"; ...</td>\n",
       "      <td>2057828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104549</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>63719980</td>\n",
       "      <td>65707226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188107\"; gene_version \"15\"; ...</td>\n",
       "      <td>1987246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041854</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65912457</td>\n",
       "      <td>67763637</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183230\"; gene_version \"18\"; ...</td>\n",
       "      <td>1851180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040306</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>53802771</td>\n",
       "      <td>55627942</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150275\"; gene_version \"20\"; ...</td>\n",
       "      <td>1825171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "1601567       8  ensembl_havana    gene    2935353    4994972     .      -   \n",
       "3030998      20  ensembl_havana    gene   13995369   16053197     .      +   \n",
       "1104549       6  ensembl_havana    gene   63719980   65707226     .      -   \n",
       "2041854      10  ensembl_havana    gene   65912457   67763637     .      -   \n",
       "2040306      10  ensembl_havana    gene   53802771   55627942     .      -   \n",
       "\n",
       "        frame                                              group  gene_length  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  \n",
       "1601567     .  gene_id \"ENSG00000183117\"; gene_version \"20\"; ...      2059619  \n",
       "3030998     .  gene_id \"ENSG00000172264\"; gene_version \"18\"; ...      2057828  \n",
       "1104549     .  gene_id \"ENSG00000188107\"; gene_version \"15\"; ...      1987246  \n",
       "2041854     .  gene_id \"ENSG00000183230\"; gene_version \"18\"; ...      1851180  \n",
       "2040306     .  gene_id \"ENSG00000150275\"; gene_version \"20\"; ...      1825171  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "longest_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c1a71c30-b778-45bb-becd-5291fc5f749a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_genes_descending = genes.sort_values(by=['gene_length'], ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "422989f2-5be2-44fa-8184-c3f6fd12e179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2913291</th>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>43131208</td>\n",
       "      <td>43131270</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000285057\"; gene_version \"1\"; g...</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118533</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>93083598</td>\n",
       "      <td>93083675</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256393\"; gene_version \"1\"; g...</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362811</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22852616</td>\n",
       "      <td>22852728</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212335\"; gene_version \"1\"; g...</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904441</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838903</td>\n",
       "      <td>90839037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187653\"; gene_version \"11\"; ...</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133423</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>73373108</td>\n",
       "      <td>73373263</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235174\"; gene_version \"1\"; g...</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19285 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2913291      17  ensembl_havana    gene   43131208   43131270     .      +   \n",
       "2118533      12  ensembl_havana    gene   93083598   93083675     .      -   \n",
       "2362811      14  ensembl_havana    gene   22852616   22852728     .      -   \n",
       "904441        4  ensembl_havana    gene   90838903   90839037     .      -   \n",
       "1133423       6  ensembl_havana    gene   73373108   73373263     .      -   \n",
       "...         ...             ...     ...        ...        ...   ...    ...   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "\n",
       "        frame                                              group  gene_length  \n",
       "2913291     .  gene_id \"ENSG00000285057\"; gene_version \"1\"; g...           62  \n",
       "2118533     .  gene_id \"ENSG00000256393\"; gene_version \"1\"; g...           77  \n",
       "2362811     .  gene_id \"ENSG00000212335\"; gene_version \"1\"; g...          112  \n",
       "904441      .  gene_id \"ENSG00000187653\"; gene_version \"11\"; ...          134  \n",
       "1133423     .  gene_id \"ENSG00000235174\"; gene_version \"1\"; g...          155  \n",
       "...       ...                                                ...          ...  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "\n",
       "[19285 rows x 10 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_genes_descending"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "66da7d80-5480-4a18-afa0-6bc428bd618e",
   "metadata": {},
   "outputs": [],
   "source": [
    "shortest_genes = sorted_genes_descending.iloc[0:20,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0b9fddd5-3f19-4b87-bca6-83774d3ed7e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2913291</th>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>43131208</td>\n",
       "      <td>43131270</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000285057\"; gene_version \"1\"; g...</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118533</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>93083598</td>\n",
       "      <td>93083675</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256393\"; gene_version \"1\"; g...</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362811</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22852616</td>\n",
       "      <td>22852728</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212335\"; gene_version \"1\"; g...</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904441</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838903</td>\n",
       "      <td>90839037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187653\"; gene_version \"11\"; ...</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133423</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>73373108</td>\n",
       "      <td>73373263</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235174\"; gene_version \"1\"; g...</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129959</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>27649419</td>\n",
       "      <td>27649610</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235912\"; gene_version \"1\"; g...</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847776</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>82689559</td>\n",
       "      <td>82689768</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227097\"; gene_version \"5\"; g...</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2938518</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>52409042</td>\n",
       "      <td>52409257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000280098\"; gene_version \"1\"; g...</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675285</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>186596976</td>\n",
       "      <td>186597200</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000283390\"; gene_version \"1\"; g...</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988497</th>\n",
       "      <td>5</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>62512246</td>\n",
       "      <td>62512482</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268942\"; gene_version \"2\"; g...</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1967063</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29697935</td>\n",
       "      <td>29698174</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225976\"; gene_version \"5\"; g...</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255568</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>18992467</td>\n",
       "      <td>18992709</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185275\"; gene_version \"6\"; g...</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363759</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30718525</td>\n",
       "      <td>30718777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000231068\"; gene_version \"1\"; g...</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2274799</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3211663</td>\n",
       "      <td>3211917</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227081\"; gene_version \"5\"; g...</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1268412</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>129410113</td>\n",
       "      <td>129410370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000238072\"; gene_version \"1\"; g...</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105987</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>152622834</td>\n",
       "      <td>152623103</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185962\"; gene_version \"1\"; g...</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2594158</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>26032539</td>\n",
       "      <td>26032811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230330\"; gene_version \"1\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2526556</th>\n",
       "      <td>15</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29730713</td>\n",
       "      <td>29730985</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234664\"; gene_version \"1\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363751</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30642864</td>\n",
       "      <td>30643136</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000206104\"; gene_version \"2\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363727</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30616425</td>\n",
       "      <td>30616699</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000244624\"; gene_version \"3\"; g...</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2913291      17  ensembl_havana    gene   43131208   43131270     .      +   \n",
       "2118533      12  ensembl_havana    gene   93083598   93083675     .      -   \n",
       "2362811      14  ensembl_havana    gene   22852616   22852728     .      -   \n",
       "904441        4  ensembl_havana    gene   90838903   90839037     .      -   \n",
       "1133423       6  ensembl_havana    gene   73373108   73373263     .      -   \n",
       "129959        1  ensembl_havana    gene   27649419   27649610     .      +   \n",
       "1847776      11  ensembl_havana    gene   82689559   82689768     .      +   \n",
       "2938518      18  ensembl_havana    gene   52409042   52409257     .      +   \n",
       "675285        3  ensembl_havana    gene  186596976  186597200     .      +   \n",
       "988497        5  ensembl_havana    gene   62512246   62512482     .      -   \n",
       "1967063      10  ensembl_havana    gene   29697935   29698174     .      +   \n",
       "3255568       Y  ensembl_havana    gene   18992467   18992709     .      -   \n",
       "3363759      21  ensembl_havana    gene   30718525   30718777     .      -   \n",
       "2274799      12  ensembl_havana    gene    3211663    3211917     .      -   \n",
       "1268412       7  ensembl_havana    gene  129410113  129410370     .      -   \n",
       "105987        1  ensembl_havana    gene  152622834  152623103     .      -   \n",
       "2594158      16  ensembl_havana    gene   26032539   26032811     .      -   \n",
       "2526556      15  ensembl_havana    gene   29730713   29730985     .      +   \n",
       "3363751      21  ensembl_havana    gene   30642864   30643136     .      +   \n",
       "3363727      21  ensembl_havana    gene   30616425   30616699     .      +   \n",
       "\n",
       "        frame                                              group  gene_length  \n",
       "2913291     .  gene_id \"ENSG00000285057\"; gene_version \"1\"; g...           62  \n",
       "2118533     .  gene_id \"ENSG00000256393\"; gene_version \"1\"; g...           77  \n",
       "2362811     .  gene_id \"ENSG00000212335\"; gene_version \"1\"; g...          112  \n",
       "904441      .  gene_id \"ENSG00000187653\"; gene_version \"11\"; ...          134  \n",
       "1133423     .  gene_id \"ENSG00000235174\"; gene_version \"1\"; g...          155  \n",
       "129959      .  gene_id \"ENSG00000235912\"; gene_version \"1\"; g...          191  \n",
       "1847776     .  gene_id \"ENSG00000227097\"; gene_version \"5\"; g...          209  \n",
       "2938518     .  gene_id \"ENSG00000280098\"; gene_version \"1\"; g...          215  \n",
       "675285      .  gene_id \"ENSG00000283390\"; gene_version \"1\"; g...          224  \n",
       "988497      .  gene_id \"ENSG00000268942\"; gene_version \"2\"; g...          236  \n",
       "1967063     .  gene_id \"ENSG00000225976\"; gene_version \"5\"; g...          239  \n",
       "3255568     .  gene_id \"ENSG00000185275\"; gene_version \"6\"; g...          242  \n",
       "3363759     .  gene_id \"ENSG00000231068\"; gene_version \"1\"; g...          252  \n",
       "2274799     .  gene_id \"ENSG00000227081\"; gene_version \"5\"; g...          254  \n",
       "1268412     .  gene_id \"ENSG00000238072\"; gene_version \"1\"; g...          257  \n",
       "105987      .  gene_id \"ENSG00000185962\"; gene_version \"1\"; g...          269  \n",
       "2594158     .  gene_id \"ENSG00000230330\"; gene_version \"1\"; g...          272  \n",
       "2526556     .  gene_id \"ENSG00000234664\"; gene_version \"1\"; g...          272  \n",
       "3363751     .  gene_id \"ENSG00000206104\"; gene_version \"2\"; g...          272  \n",
       "3363727     .  gene_id \"ENSG00000244624\"; gene_version \"3\"; g...          274  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shortest_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e31ba29d-d7d9-4dd1-844e-7629aa6eb4cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_group = genes[\"group\"].str.split(\";\").to_frame().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c8168664-4aed-4662-8dbb-04deb48543db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[gene_id \"ENSG00000160072\",  gene_version \"20\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>[gene_id \"ENSG00000142611\",  gene_version \"17\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>[gene_id \"ENSG00000157911\",  gene_version \"11\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>[gene_id \"ENSG00000157933\",  gene_version \"10\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>[gene_id \"ENSG00000142655\",  gene_version \"13\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 group\n",
       "0    [gene_id \"ENSG00000160072\",  gene_version \"20\"...\n",
       "132  [gene_id \"ENSG00000142611\",  gene_version \"17\"...\n",
       "361  [gene_id \"ENSG00000157911\",  gene_version \"11\"...\n",
       "517  [gene_id \"ENSG00000157933\",  gene_version \"10\"...\n",
       "580  [gene_id \"ENSG00000142655\",  gene_version \"13\"..."
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2cba2c7d-5e88-434e-a255-31bea590bb02",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'list'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [51]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mens_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\u001b[43msplit_group\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgroup\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43minner\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/frame.py:9345\u001b[0m, in \u001b[0;36mDataFrame.merge\u001b[0;34m(self, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m   9326\u001b[0m \u001b[38;5;129m@Substitution\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   9327\u001b[0m \u001b[38;5;129m@Appender\u001b[39m(_merge_doc, indents\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m)\n\u001b[1;32m   9328\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmerge\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9341\u001b[0m     validate: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   9342\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame:\n\u001b[1;32m   9343\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mreshape\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmerge\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m merge\n\u001b[0;32m-> 9345\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   9346\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9347\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9348\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhow\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9349\u001b[0m \u001b[43m        \u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9350\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_on\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mleft_on\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9351\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_on\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mright_on\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9352\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mleft_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9353\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mright_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9354\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9355\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuffixes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msuffixes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9356\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9357\u001b[0m \u001b[43m        \u001b[49m\u001b[43mindicator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindicator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9358\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9359\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:122\u001b[0m, in \u001b[0;36mmerge\u001b[0;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;129m@Substitution\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mleft : DataFrame or named Series\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     91\u001b[0m \u001b[38;5;129m@Appender\u001b[39m(_merge_doc, indents\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmerge\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    105\u001b[0m     validate: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    106\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame:\n\u001b[1;32m    107\u001b[0m     op \u001b[38;5;241m=\u001b[39m _MergeOperation(\n\u001b[1;32m    108\u001b[0m         left,\n\u001b[1;32m    109\u001b[0m         right,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    120\u001b[0m         validate\u001b[38;5;241m=\u001b[39mvalidate,\n\u001b[1;32m    121\u001b[0m     )\n\u001b[0;32m--> 122\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:716\u001b[0m, in \u001b[0;36m_MergeOperation.get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    713\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindicator:\n\u001b[1;32m    714\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_indicator_pre_merge(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright)\n\u001b[0;32m--> 716\u001b[0m join_index, left_indexer, right_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_join_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    718\u001b[0m llabels, rlabels \u001b[38;5;241m=\u001b[39m _items_overlap_with_suffix(\n\u001b[1;32m    719\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft\u001b[38;5;241m.\u001b[39m_info_axis, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright\u001b[38;5;241m.\u001b[39m_info_axis, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msuffixes\n\u001b[1;32m    720\u001b[0m )\n\u001b[1;32m    722\u001b[0m lindexers \u001b[38;5;241m=\u001b[39m {\u001b[38;5;241m1\u001b[39m: left_indexer} \u001b[38;5;28;01mif\u001b[39;00m left_indexer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m {}\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:967\u001b[0m, in \u001b[0;36m_MergeOperation._get_join_info\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    963\u001b[0m     join_index, right_indexer, left_indexer \u001b[38;5;241m=\u001b[39m _left_join_on_index(\n\u001b[1;32m    964\u001b[0m         right_ax, left_ax, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright_join_keys, sort\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msort\n\u001b[1;32m    965\u001b[0m     )\n\u001b[1;32m    966\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 967\u001b[0m     (left_indexer, right_indexer) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_join_indexers\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    969\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright_index:\n\u001b[1;32m    970\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:941\u001b[0m, in \u001b[0;36m_MergeOperation._get_join_indexers\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    939\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_join_indexers\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mtuple\u001b[39m[npt\u001b[38;5;241m.\u001b[39mNDArray[np\u001b[38;5;241m.\u001b[39mintp], npt\u001b[38;5;241m.\u001b[39mNDArray[np\u001b[38;5;241m.\u001b[39mintp]]:\n\u001b[1;32m    940\u001b[0m     \u001b[38;5;124;03m\"\"\"return the join indexers\"\"\"\u001b[39;00m\n\u001b[0;32m--> 941\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mget_join_indexers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    942\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mleft_join_keys\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mright_join_keys\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhow\u001b[49m\n\u001b[1;32m    943\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:1484\u001b[0m, in \u001b[0;36mget_join_indexers\u001b[0;34m(left_keys, right_keys, sort, how, **kwargs)\u001b[0m\n\u001b[1;32m   1479\u001b[0m \u001b[38;5;66;03m# get left & right join labels and num. of levels at each location\u001b[39;00m\n\u001b[1;32m   1480\u001b[0m mapped \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m   1481\u001b[0m     _factorize_keys(left_keys[n], right_keys[n], sort\u001b[38;5;241m=\u001b[39msort, how\u001b[38;5;241m=\u001b[39mhow)\n\u001b[1;32m   1482\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(left_keys))\n\u001b[1;32m   1483\u001b[0m )\n\u001b[0;32m-> 1484\u001b[0m zipped \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mzip\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmapped\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1485\u001b[0m llab, rlab, shape \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mlist\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m zipped)\n\u001b[1;32m   1487\u001b[0m \u001b[38;5;66;03m# get flat i8 keys from label lists\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:1481\u001b[0m, in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1475\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(left_keys) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(\n\u001b[1;32m   1476\u001b[0m     right_keys\n\u001b[1;32m   1477\u001b[0m ), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mleft_key and right_keys must be the same length\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1479\u001b[0m \u001b[38;5;66;03m# get left & right join labels and num. of levels at each location\u001b[39;00m\n\u001b[1;32m   1480\u001b[0m mapped \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m-> 1481\u001b[0m     \u001b[43m_factorize_keys\u001b[49m\u001b[43m(\u001b[49m\u001b[43mleft_keys\u001b[49m\u001b[43m[\u001b[49m\u001b[43mn\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mright_keys\u001b[49m\u001b[43m[\u001b[49m\u001b[43mn\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhow\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1482\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(left_keys))\n\u001b[1;32m   1483\u001b[0m )\n\u001b[1;32m   1484\u001b[0m zipped \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;241m*\u001b[39mmapped)\n\u001b[1;32m   1485\u001b[0m llab, rlab, shape \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mlist\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m zipped)\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py:2187\u001b[0m, in \u001b[0;36m_factorize_keys\u001b[0;34m(lk, rk, sort, how)\u001b[0m\n\u001b[1;32m   2183\u001b[0m llab \u001b[38;5;241m=\u001b[39m rizer\u001b[38;5;241m.\u001b[39mfactorize(lk)  \u001b[38;5;66;03m# type: ignore[arg-type]\u001b[39;00m\n\u001b[1;32m   2184\u001b[0m \u001b[38;5;66;03m# Argument 1 to \"factorize\" of \"ObjectFactorizer\" has incompatible type\u001b[39;00m\n\u001b[1;32m   2185\u001b[0m \u001b[38;5;66;03m# \"Union[ndarray[Any, dtype[signedinteger[_64Bit]]],\u001b[39;00m\n\u001b[1;32m   2186\u001b[0m \u001b[38;5;66;03m# ndarray[Any, dtype[object_]]]\"; expected \"ndarray[Any, dtype[object_]]\"\u001b[39;00m\n\u001b[0;32m-> 2187\u001b[0m rlab \u001b[38;5;241m=\u001b[39m \u001b[43mrizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfactorize\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrk\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# type: ignore[arg-type]\u001b[39;00m\n\u001b[1;32m   2188\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m llab\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mdtype(np\u001b[38;5;241m.\u001b[39mintp), llab\u001b[38;5;241m.\u001b[39mdtype\n\u001b[1;32m   2189\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m rlab\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mdtype(np\u001b[38;5;241m.\u001b[39mintp), rlab\u001b[38;5;241m.\u001b[39mdtype\n",
      "File \u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_libs/hashtable.pyx:124\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.ObjectFactorizer.factorize\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5402\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_labels\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5310\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable._unique\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: unhashable type: 'list'"
     ]
    }
   ],
   "source": [
    "ens_df.merge(split_group, on=\"group\", how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5c373f69-1dda-425d-886f-2ecadfa96cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_per_chr = genes.groupby(\"seqname\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b0a2c6bb-d2ad-42ef-9218-af113658a343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f90bc480a00>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_per_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c6393bd3-ce85-487e-865c-d303f25f0d2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">start</th>\n",
       "      <th colspan=\"5\" halign=\"left\">end</th>\n",
       "      <th colspan=\"8\" halign=\"left\">gene_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>...</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seqname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2009.0</td>\n",
       "      <td>1.128017e+08</td>\n",
       "      <td>7.632351e+07</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>36466043.00</td>\n",
       "      <td>113390515.0</td>\n",
       "      <td>1.691325e+08</td>\n",
       "      <td>248906196.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>1.128589e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>1.693679e+08</td>\n",
       "      <td>248919946.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>57141.112992</td>\n",
       "      <td>107043.255133</td>\n",
       "      <td>191.0</td>\n",
       "      <td>9072.00</td>\n",
       "      <td>24448.0</td>\n",
       "      <td>58905.00</td>\n",
       "      <td>1551956.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1222.0</td>\n",
       "      <td>1.249662e+08</td>\n",
       "      <td>7.387081e+07</td>\n",
       "      <td>38814.0</td>\n",
       "      <td>64182981.25</td>\n",
       "      <td>113083642.5</td>\n",
       "      <td>1.975551e+08</td>\n",
       "      <td>241869600.0</td>\n",
       "      <td>1222.0</td>\n",
       "      <td>1.250528e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>1.976455e+08</td>\n",
       "      <td>241873823.0</td>\n",
       "      <td>1222.0</td>\n",
       "      <td>86600.607201</td>\n",
       "      <td>139473.259443</td>\n",
       "      <td>313.0</td>\n",
       "      <td>14108.50</td>\n",
       "      <td>43435.0</td>\n",
       "      <td>101049.25</td>\n",
       "      <td>1463630.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1040.0</td>\n",
       "      <td>9.710336e+07</td>\n",
       "      <td>5.892302e+07</td>\n",
       "      <td>196763.0</td>\n",
       "      <td>47007237.50</td>\n",
       "      <td>100447075.5</td>\n",
       "      <td>1.461002e+08</td>\n",
       "      <td>197960200.0</td>\n",
       "      <td>1040.0</td>\n",
       "      <td>9.719733e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.461855e+08</td>\n",
       "      <td>198043720.0</td>\n",
       "      <td>1040.0</td>\n",
       "      <td>93965.008654</td>\n",
       "      <td>171439.552392</td>\n",
       "      <td>224.0</td>\n",
       "      <td>14727.50</td>\n",
       "      <td>41582.5</td>\n",
       "      <td>93488.50</td>\n",
       "      <td>1743269.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>736.0</td>\n",
       "      <td>8.588778e+07</td>\n",
       "      <td>5.733972e+07</td>\n",
       "      <td>53286.0</td>\n",
       "      <td>39488817.25</td>\n",
       "      <td>80301184.5</td>\n",
       "      <td>1.376802e+08</td>\n",
       "      <td>190173774.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>8.598419e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.377100e+08</td>\n",
       "      <td>190185942.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>96408.373641</td>\n",
       "      <td>154822.552760</td>\n",
       "      <td>134.0</td>\n",
       "      <td>16185.75</td>\n",
       "      <td>48647.0</td>\n",
       "      <td>112178.75</td>\n",
       "      <td>1506191.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>835.0</td>\n",
       "      <td>1.096735e+08</td>\n",
       "      <td>5.281069e+07</td>\n",
       "      <td>92168.0</td>\n",
       "      <td>69178354.50</td>\n",
       "      <td>129748094.0</td>\n",
       "      <td>1.502579e+08</td>\n",
       "      <td>181367268.0</td>\n",
       "      <td>835.0</td>\n",
       "      <td>1.097591e+08</td>\n",
       "      <td>...</td>\n",
       "      <td>1.503147e+08</td>\n",
       "      <td>181368262.0</td>\n",
       "      <td>835.0</td>\n",
       "      <td>85639.657485</td>\n",
       "      <td>128743.813319</td>\n",
       "      <td>236.0</td>\n",
       "      <td>14679.00</td>\n",
       "      <td>43041.0</td>\n",
       "      <td>102746.50</td>\n",
       "      <td>1553082.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>995.0</td>\n",
       "      <td>6.950979e+07</td>\n",
       "      <td>4.953143e+07</td>\n",
       "      <td>131910.0</td>\n",
       "      <td>31150493.00</td>\n",
       "      <td>44000580.0</td>\n",
       "      <td>1.131023e+08</td>\n",
       "      <td>170575295.0</td>\n",
       "      <td>995.0</td>\n",
       "      <td>6.958302e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.131074e+08</td>\n",
       "      <td>170584692.0</td>\n",
       "      <td>995.0</td>\n",
       "      <td>73231.225126</td>\n",
       "      <td>136261.985072</td>\n",
       "      <td>155.0</td>\n",
       "      <td>9442.00</td>\n",
       "      <td>25999.0</td>\n",
       "      <td>79145.50</td>\n",
       "      <td>1987246.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>804.0</td>\n",
       "      <td>8.459091e+07</td>\n",
       "      <td>4.757134e+07</td>\n",
       "      <td>192571.0</td>\n",
       "      <td>43067565.75</td>\n",
       "      <td>95678002.5</td>\n",
       "      <td>1.284611e+08</td>\n",
       "      <td>159028175.0</td>\n",
       "      <td>804.0</td>\n",
       "      <td>8.467858e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.284705e+08</td>\n",
       "      <td>159144867.0</td>\n",
       "      <td>804.0</td>\n",
       "      <td>87677.218905</td>\n",
       "      <td>167218.151568</td>\n",
       "      <td>257.0</td>\n",
       "      <td>12206.00</td>\n",
       "      <td>32601.0</td>\n",
       "      <td>86587.50</td>\n",
       "      <td>2304996.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>304.0</td>\n",
       "      <td>6.481184e+07</td>\n",
       "      <td>4.872622e+07</td>\n",
       "      <td>166086.0</td>\n",
       "      <td>22344287.50</td>\n",
       "      <td>42356258.5</td>\n",
       "      <td>1.044369e+08</td>\n",
       "      <td>144930358.0</td>\n",
       "      <td>304.0</td>\n",
       "      <td>6.489234e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.044976e+08</td>\n",
       "      <td>144950888.0</td>\n",
       "      <td>304.0</td>\n",
       "      <td>80495.348684</td>\n",
       "      <td>178037.347733</td>\n",
       "      <td>662.0</td>\n",
       "      <td>9758.25</td>\n",
       "      <td>29936.0</td>\n",
       "      <td>77911.25</td>\n",
       "      <td>2059619.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>752.0</td>\n",
       "      <td>9.028893e+07</td>\n",
       "      <td>4.346695e+07</td>\n",
       "      <td>14475.0</td>\n",
       "      <td>38600497.25</td>\n",
       "      <td>104299261.5</td>\n",
       "      <td>1.283815e+08</td>\n",
       "      <td>137877782.0</td>\n",
       "      <td>752.0</td>\n",
       "      <td>9.036051e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.284033e+08</td>\n",
       "      <td>138124624.0</td>\n",
       "      <td>752.0</td>\n",
       "      <td>71571.687500</td>\n",
       "      <td>140419.868119</td>\n",
       "      <td>341.0</td>\n",
       "      <td>7955.00</td>\n",
       "      <td>27983.0</td>\n",
       "      <td>84100.75</td>\n",
       "      <td>2298756.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>725.0</td>\n",
       "      <td>7.515648e+07</td>\n",
       "      <td>3.793313e+07</td>\n",
       "      <td>46888.0</td>\n",
       "      <td>45972489.00</td>\n",
       "      <td>80271820.0</td>\n",
       "      <td>1.026445e+08</td>\n",
       "      <td>133683668.0</td>\n",
       "      <td>725.0</td>\n",
       "      <td>7.524120e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.026583e+08</td>\n",
       "      <td>133684936.0</td>\n",
       "      <td>725.0</td>\n",
       "      <td>84721.099310</td>\n",
       "      <td>155353.760367</td>\n",
       "      <td>239.0</td>\n",
       "      <td>14789.00</td>\n",
       "      <td>38975.0</td>\n",
       "      <td>90568.00</td>\n",
       "      <td>1851180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1261.0</td>\n",
       "      <td>6.147474e+07</td>\n",
       "      <td>3.922904e+07</td>\n",
       "      <td>112967.0</td>\n",
       "      <td>28108248.00</td>\n",
       "      <td>63079940.0</td>\n",
       "      <td>8.971030e+07</td>\n",
       "      <td>134378504.0</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>6.152638e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>8.971787e+07</td>\n",
       "      <td>134412242.0</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>51645.394132</td>\n",
       "      <td>123045.798674</td>\n",
       "      <td>209.0</td>\n",
       "      <td>6022.00</td>\n",
       "      <td>16346.0</td>\n",
       "      <td>47649.00</td>\n",
       "      <td>2172910.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>998.0</td>\n",
       "      <td>6.280751e+07</td>\n",
       "      <td>4.025189e+07</td>\n",
       "      <td>66767.0</td>\n",
       "      <td>27361311.00</td>\n",
       "      <td>55832520.5</td>\n",
       "      <td>1.029275e+08</td>\n",
       "      <td>133218312.0</td>\n",
       "      <td>998.0</td>\n",
       "      <td>6.287369e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.029600e+08</td>\n",
       "      <td>133235877.0</td>\n",
       "      <td>998.0</td>\n",
       "      <td>66181.471944</td>\n",
       "      <td>109277.089171</td>\n",
       "      <td>77.0</td>\n",
       "      <td>10998.75</td>\n",
       "      <td>29043.5</td>\n",
       "      <td>71513.00</td>\n",
       "      <td>1258479.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>319.0</td>\n",
       "      <td>5.972681e+07</td>\n",
       "      <td>3.144025e+07</td>\n",
       "      <td>18408806.0</td>\n",
       "      <td>33059780.00</td>\n",
       "      <td>49247925.0</td>\n",
       "      <td>9.486473e+07</td>\n",
       "      <td>114314482.0</td>\n",
       "      <td>319.0</td>\n",
       "      <td>5.983173e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>9.500701e+07</td>\n",
       "      <td>114337626.0</td>\n",
       "      <td>319.0</td>\n",
       "      <td>104916.489028</td>\n",
       "      <td>171387.751752</td>\n",
       "      <td>398.0</td>\n",
       "      <td>19597.50</td>\n",
       "      <td>46371.0</td>\n",
       "      <td>112931.50</td>\n",
       "      <td>1475061.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>616.0</td>\n",
       "      <td>6.112675e+07</td>\n",
       "      <td>2.932076e+07</td>\n",
       "      <td>18601117.0</td>\n",
       "      <td>30313488.25</td>\n",
       "      <td>63923065.5</td>\n",
       "      <td>9.011063e+07</td>\n",
       "      <td>106874583.0</td>\n",
       "      <td>616.0</td>\n",
       "      <td>6.119391e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>9.020825e+07</td>\n",
       "      <td>106875071.0</td>\n",
       "      <td>616.0</td>\n",
       "      <td>67166.748377</td>\n",
       "      <td>125914.150080</td>\n",
       "      <td>112.0</td>\n",
       "      <td>8988.75</td>\n",
       "      <td>25254.0</td>\n",
       "      <td>77939.25</td>\n",
       "      <td>1697918.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>577.0</td>\n",
       "      <td>6.066787e+07</td>\n",
       "      <td>2.111246e+07</td>\n",
       "      <td>20562375.0</td>\n",
       "      <td>42348103.00</td>\n",
       "      <td>63321378.0</td>\n",
       "      <td>7.710066e+07</td>\n",
       "      <td>101922042.0</td>\n",
       "      <td>577.0</td>\n",
       "      <td>6.074582e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>7.742014e+07</td>\n",
       "      <td>101923113.0</td>\n",
       "      <td>577.0</td>\n",
       "      <td>77954.305026</td>\n",
       "      <td>111304.550321</td>\n",
       "      <td>272.0</td>\n",
       "      <td>15130.00</td>\n",
       "      <td>38526.0</td>\n",
       "      <td>97223.00</td>\n",
       "      <td>969298.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>823.0</td>\n",
       "      <td>3.951028e+07</td>\n",
       "      <td>3.014854e+07</td>\n",
       "      <td>46407.0</td>\n",
       "      <td>11097079.00</td>\n",
       "      <td>30698209.0</td>\n",
       "      <td>6.788907e+07</td>\n",
       "      <td>90173217.0</td>\n",
       "      <td>823.0</td>\n",
       "      <td>3.955786e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>6.790927e+07</td>\n",
       "      <td>90222851.0</td>\n",
       "      <td>823.0</td>\n",
       "      <td>47577.639125</td>\n",
       "      <td>118305.706859</td>\n",
       "      <td>272.0</td>\n",
       "      <td>7688.00</td>\n",
       "      <td>19768.0</td>\n",
       "      <td>51933.50</td>\n",
       "      <td>2473538.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1138.0</td>\n",
       "      <td>4.070888e+07</td>\n",
       "      <td>2.490666e+07</td>\n",
       "      <td>137569.0</td>\n",
       "      <td>18230430.00</td>\n",
       "      <td>41311306.0</td>\n",
       "      <td>5.988465e+07</td>\n",
       "      <td>83079609.0</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>4.074930e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>5.993617e+07</td>\n",
       "      <td>83095122.0</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>40418.720562</td>\n",
       "      <td>75646.899374</td>\n",
       "      <td>62.0</td>\n",
       "      <td>7108.00</td>\n",
       "      <td>17021.5</td>\n",
       "      <td>43190.75</td>\n",
       "      <td>1161877.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>265.0</td>\n",
       "      <td>3.948244e+07</td>\n",
       "      <td>2.395232e+07</td>\n",
       "      <td>47221.0</td>\n",
       "      <td>15056845.00</td>\n",
       "      <td>37243040.0</td>\n",
       "      <td>5.926704e+07</td>\n",
       "      <td>80157232.0</td>\n",
       "      <td>265.0</td>\n",
       "      <td>3.959190e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>5.927409e+07</td>\n",
       "      <td>80247514.0</td>\n",
       "      <td>265.0</td>\n",
       "      <td>109459.656604</td>\n",
       "      <td>148031.437933</td>\n",
       "      <td>215.0</td>\n",
       "      <td>26960.00</td>\n",
       "      <td>52441.0</td>\n",
       "      <td>137560.00</td>\n",
       "      <td>1195706.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1122.0</td>\n",
       "      <td>3.028500e+07</td>\n",
       "      <td>1.930766e+07</td>\n",
       "      <td>107104.0</td>\n",
       "      <td>11543918.75</td>\n",
       "      <td>35746980.5</td>\n",
       "      <td>4.883108e+07</td>\n",
       "      <td>58561931.0</td>\n",
       "      <td>1122.0</td>\n",
       "      <td>3.030948e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.886059e+07</td>\n",
       "      <td>58573575.0</td>\n",
       "      <td>1122.0</td>\n",
       "      <td>24486.719251</td>\n",
       "      <td>33383.796074</td>\n",
       "      <td>741.0</td>\n",
       "      <td>7882.25</td>\n",
       "      <td>15428.0</td>\n",
       "      <td>29851.25</td>\n",
       "      <td>485248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>536.0</td>\n",
       "      <td>3.466741e+07</td>\n",
       "      <td>1.990573e+07</td>\n",
       "      <td>87250.0</td>\n",
       "      <td>18812769.25</td>\n",
       "      <td>36271233.5</td>\n",
       "      <td>4.928198e+07</td>\n",
       "      <td>64290385.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>3.472166e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.934577e+07</td>\n",
       "      <td>64313132.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>54251.886194</td>\n",
       "      <td>125214.452212</td>\n",
       "      <td>305.0</td>\n",
       "      <td>9654.25</td>\n",
       "      <td>23600.5</td>\n",
       "      <td>57606.00</td>\n",
       "      <td>2057828.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>70.0</td>\n",
       "      <td>2.901048e+07</td>\n",
       "      <td>1.007805e+07</td>\n",
       "      <td>5553637.0</td>\n",
       "      <td>25599056.50</td>\n",
       "      <td>30980072.5</td>\n",
       "      <td>3.611879e+07</td>\n",
       "      <td>46598604.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>2.910594e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.612821e+07</td>\n",
       "      <td>46605208.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>95460.300000</td>\n",
       "      <td>177068.179267</td>\n",
       "      <td>369.0</td>\n",
       "      <td>15188.75</td>\n",
       "      <td>38115.5</td>\n",
       "      <td>87553.00</td>\n",
       "      <td>1216866.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>361.0</td>\n",
       "      <td>3.244761e+07</td>\n",
       "      <td>9.997649e+06</td>\n",
       "      <td>15528192.0</td>\n",
       "      <td>23179704.00</td>\n",
       "      <td>31618491.0</td>\n",
       "      <td>3.952909e+07</td>\n",
       "      <td>50767501.0</td>\n",
       "      <td>361.0</td>\n",
       "      <td>3.248780e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.953276e+07</td>\n",
       "      <td>50801309.0</td>\n",
       "      <td>361.0</td>\n",
       "      <td>40188.700831</td>\n",
       "      <td>70611.924513</td>\n",
       "      <td>476.0</td>\n",
       "      <td>9252.00</td>\n",
       "      <td>20297.0</td>\n",
       "      <td>41999.00</td>\n",
       "      <td>760615.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>240.0</td>\n",
       "      <td>3.389139e+07</td>\n",
       "      <td>1.828087e+07</td>\n",
       "      <td>812488.0</td>\n",
       "      <td>17783050.50</td>\n",
       "      <td>40251903.0</td>\n",
       "      <td>4.806474e+07</td>\n",
       "      <td>57977076.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>3.391647e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.812576e+07</td>\n",
       "      <td>58003349.0</td>\n",
       "      <td>240.0</td>\n",
       "      <td>25084.708333</td>\n",
       "      <td>24626.483113</td>\n",
       "      <td>838.0</td>\n",
       "      <td>8399.25</td>\n",
       "      <td>16572.5</td>\n",
       "      <td>33944.25</td>\n",
       "      <td>144080.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>147.0</td>\n",
       "      <td>3.477356e+07</td>\n",
       "      <td>1.194907e+07</td>\n",
       "      <td>5022531.0</td>\n",
       "      <td>30765611.00</td>\n",
       "      <td>39342315.0</td>\n",
       "      <td>4.382737e+07</td>\n",
       "      <td>46635595.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>3.482753e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.389644e+07</td>\n",
       "      <td>46665124.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>53972.210884</td>\n",
       "      <td>90592.560662</td>\n",
       "      <td>252.0</td>\n",
       "      <td>8047.00</td>\n",
       "      <td>27325.0</td>\n",
       "      <td>65406.00</td>\n",
       "      <td>836159.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>91.0</td>\n",
       "      <td>3.752528e+07</td>\n",
       "      <td>8.779680e+06</td>\n",
       "      <td>16590751.0</td>\n",
       "      <td>31303699.00</td>\n",
       "      <td>39743044.0</td>\n",
       "      <td>4.479170e+07</td>\n",
       "      <td>50343304.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>3.759145e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.493636e+07</td>\n",
       "      <td>50445090.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>66167.461538</td>\n",
       "      <td>89383.026598</td>\n",
       "      <td>1379.0</td>\n",
       "      <td>11442.00</td>\n",
       "      <td>32987.0</td>\n",
       "      <td>83496.50</td>\n",
       "      <td>550561.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>71.0</td>\n",
       "      <td>7.743717e+07</td>\n",
       "      <td>4.532863e+07</td>\n",
       "      <td>497258.0</td>\n",
       "      <td>38360688.00</td>\n",
       "      <td>87322943.0</td>\n",
       "      <td>1.075721e+08</td>\n",
       "      <td>156388922.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>7.753937e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.076262e+08</td>\n",
       "      <td>156640654.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>102202.507042</td>\n",
       "      <td>181652.537774</td>\n",
       "      <td>650.0</td>\n",
       "      <td>18552.00</td>\n",
       "      <td>39586.0</td>\n",
       "      <td>113986.00</td>\n",
       "      <td>1195210.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>356.0</td>\n",
       "      <td>7.896998e+07</td>\n",
       "      <td>4.495036e+07</td>\n",
       "      <td>406428.0</td>\n",
       "      <td>40294064.75</td>\n",
       "      <td>78591214.5</td>\n",
       "      <td>1.203969e+08</td>\n",
       "      <td>145052465.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>7.906666e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.204696e+08</td>\n",
       "      <td>145066685.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>96684.362360</td>\n",
       "      <td>150044.306593</td>\n",
       "      <td>347.0</td>\n",
       "      <td>13074.75</td>\n",
       "      <td>44002.5</td>\n",
       "      <td>116458.75</td>\n",
       "      <td>1214011.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X</th>\n",
       "      <td>811.0</td>\n",
       "      <td>8.470459e+07</td>\n",
       "      <td>4.792873e+07</td>\n",
       "      <td>276322.0</td>\n",
       "      <td>48067481.00</td>\n",
       "      <td>78945332.0</td>\n",
       "      <td>1.298547e+08</td>\n",
       "      <td>156020826.0</td>\n",
       "      <td>811.0</td>\n",
       "      <td>8.477032e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.298868e+08</td>\n",
       "      <td>156025710.0</td>\n",
       "      <td>811.0</td>\n",
       "      <td>65726.053021</td>\n",
       "      <td>145339.541124</td>\n",
       "      <td>347.0</td>\n",
       "      <td>5930.00</td>\n",
       "      <td>19277.0</td>\n",
       "      <td>66260.00</td>\n",
       "      <td>2241932.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>61.0</td>\n",
       "      <td>1.680047e+07</td>\n",
       "      <td>7.230866e+06</td>\n",
       "      <td>2786855.0</td>\n",
       "      <td>9527880.00</td>\n",
       "      <td>18772706.0</td>\n",
       "      <td>2.297382e+07</td>\n",
       "      <td>25975549.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1.685450e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>2.300546e+07</td>\n",
       "      <td>25988069.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>54031.918033</td>\n",
       "      <td>117153.357329</td>\n",
       "      <td>242.0</td>\n",
       "      <td>2814.00</td>\n",
       "      <td>14471.0</td>\n",
       "      <td>42272.00</td>\n",
       "      <td>741998.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          start                                                       \\\n",
       "          count          mean           std         min          25%   \n",
       "seqname                                                                \n",
       "1        2009.0  1.128017e+08  7.632351e+07     65419.0  36466043.00   \n",
       "2        1222.0  1.249662e+08  7.387081e+07     38814.0  64182981.25   \n",
       "3        1040.0  9.710336e+07  5.892302e+07    196763.0  47007237.50   \n",
       "4         736.0  8.588778e+07  5.733972e+07     53286.0  39488817.25   \n",
       "5         835.0  1.096735e+08  5.281069e+07     92168.0  69178354.50   \n",
       "6         995.0  6.950979e+07  4.953143e+07    131910.0  31150493.00   \n",
       "7         804.0  8.459091e+07  4.757134e+07    192571.0  43067565.75   \n",
       "8         304.0  6.481184e+07  4.872622e+07    166086.0  22344287.50   \n",
       "9         752.0  9.028893e+07  4.346695e+07     14475.0  38600497.25   \n",
       "10        725.0  7.515648e+07  3.793313e+07     46888.0  45972489.00   \n",
       "11       1261.0  6.147474e+07  3.922904e+07    112967.0  28108248.00   \n",
       "12        998.0  6.280751e+07  4.025189e+07     66767.0  27361311.00   \n",
       "13        319.0  5.972681e+07  3.144025e+07  18408806.0  33059780.00   \n",
       "14        616.0  6.112675e+07  2.932076e+07  18601117.0  30313488.25   \n",
       "15        577.0  6.066787e+07  2.111246e+07  20562375.0  42348103.00   \n",
       "16        823.0  3.951028e+07  3.014854e+07     46407.0  11097079.00   \n",
       "17       1138.0  4.070888e+07  2.490666e+07    137569.0  18230430.00   \n",
       "18        265.0  3.948244e+07  2.395232e+07     47221.0  15056845.00   \n",
       "19       1122.0  3.028500e+07  1.930766e+07    107104.0  11543918.75   \n",
       "20        536.0  3.466741e+07  1.990573e+07     87250.0  18812769.25   \n",
       "21         70.0  2.901048e+07  1.007805e+07   5553637.0  25599056.50   \n",
       "22        361.0  3.244761e+07  9.997649e+06  15528192.0  23179704.00   \n",
       "19        240.0  3.389139e+07  1.828087e+07    812488.0  17783050.50   \n",
       "21        147.0  3.477356e+07  1.194907e+07   5022531.0  30765611.00   \n",
       "22         91.0  3.752528e+07  8.779680e+06  16590751.0  31303699.00   \n",
       "7          71.0  7.743717e+07  4.532863e+07    497258.0  38360688.00   \n",
       "8         356.0  7.896998e+07  4.495036e+07    406428.0  40294064.75   \n",
       "X         811.0  8.470459e+07  4.792873e+07    276322.0  48067481.00   \n",
       "Y          61.0  1.680047e+07  7.230866e+06   2786855.0   9527880.00   \n",
       "\n",
       "                                                    end                ...  \\\n",
       "                 50%           75%          max   count          mean  ...   \n",
       "seqname                                                                ...   \n",
       "1        113390515.0  1.691325e+08  248906196.0  2009.0  1.128589e+08  ...   \n",
       "2        113083642.5  1.975551e+08  241869600.0  1222.0  1.250528e+08  ...   \n",
       "3        100447075.5  1.461002e+08  197960200.0  1040.0  9.719733e+07  ...   \n",
       "4         80301184.5  1.376802e+08  190173774.0   736.0  8.598419e+07  ...   \n",
       "5        129748094.0  1.502579e+08  181367268.0   835.0  1.097591e+08  ...   \n",
       "6         44000580.0  1.131023e+08  170575295.0   995.0  6.958302e+07  ...   \n",
       "7         95678002.5  1.284611e+08  159028175.0   804.0  8.467858e+07  ...   \n",
       "8         42356258.5  1.044369e+08  144930358.0   304.0  6.489234e+07  ...   \n",
       "9        104299261.5  1.283815e+08  137877782.0   752.0  9.036051e+07  ...   \n",
       "10        80271820.0  1.026445e+08  133683668.0   725.0  7.524120e+07  ...   \n",
       "11        63079940.0  8.971030e+07  134378504.0  1261.0  6.152638e+07  ...   \n",
       "12        55832520.5  1.029275e+08  133218312.0   998.0  6.287369e+07  ...   \n",
       "13        49247925.0  9.486473e+07  114314482.0   319.0  5.983173e+07  ...   \n",
       "14        63923065.5  9.011063e+07  106874583.0   616.0  6.119391e+07  ...   \n",
       "15        63321378.0  7.710066e+07  101922042.0   577.0  6.074582e+07  ...   \n",
       "16        30698209.0  6.788907e+07   90173217.0   823.0  3.955786e+07  ...   \n",
       "17        41311306.0  5.988465e+07   83079609.0  1138.0  4.074930e+07  ...   \n",
       "18        37243040.0  5.926704e+07   80157232.0   265.0  3.959190e+07  ...   \n",
       "19        35746980.5  4.883108e+07   58561931.0  1122.0  3.030948e+07  ...   \n",
       "20        36271233.5  4.928198e+07   64290385.0   536.0  3.472166e+07  ...   \n",
       "21        30980072.5  3.611879e+07   46598604.0    70.0  2.910594e+07  ...   \n",
       "22        31618491.0  3.952909e+07   50767501.0   361.0  3.248780e+07  ...   \n",
       "19        40251903.0  4.806474e+07   57977076.0   240.0  3.391647e+07  ...   \n",
       "21        39342315.0  4.382737e+07   46635595.0   147.0  3.482753e+07  ...   \n",
       "22        39743044.0  4.479170e+07   50343304.0    91.0  3.759145e+07  ...   \n",
       "7         87322943.0  1.075721e+08  156388922.0    71.0  7.753937e+07  ...   \n",
       "8         78591214.5  1.203969e+08  145052465.0   356.0  7.906666e+07  ...   \n",
       "X         78945332.0  1.298547e+08  156020826.0   811.0  8.477032e+07  ...   \n",
       "Y         18772706.0  2.297382e+07   25975549.0    61.0  1.685450e+07  ...   \n",
       "\n",
       "                                   gene_length                                \\\n",
       "                  75%          max       count           mean            std   \n",
       "seqname                                                                        \n",
       "1        1.693679e+08  248919946.0      2009.0   57141.112992  107043.255133   \n",
       "2        1.976455e+08  241873823.0      1222.0   86600.607201  139473.259443   \n",
       "3        1.461855e+08  198043720.0      1040.0   93965.008654  171439.552392   \n",
       "4        1.377100e+08  190185942.0       736.0   96408.373641  154822.552760   \n",
       "5        1.503147e+08  181368262.0       835.0   85639.657485  128743.813319   \n",
       "6        1.131074e+08  170584692.0       995.0   73231.225126  136261.985072   \n",
       "7        1.284705e+08  159144867.0       804.0   87677.218905  167218.151568   \n",
       "8        1.044976e+08  144950888.0       304.0   80495.348684  178037.347733   \n",
       "9        1.284033e+08  138124624.0       752.0   71571.687500  140419.868119   \n",
       "10       1.026583e+08  133684936.0       725.0   84721.099310  155353.760367   \n",
       "11       8.971787e+07  134412242.0      1261.0   51645.394132  123045.798674   \n",
       "12       1.029600e+08  133235877.0       998.0   66181.471944  109277.089171   \n",
       "13       9.500701e+07  114337626.0       319.0  104916.489028  171387.751752   \n",
       "14       9.020825e+07  106875071.0       616.0   67166.748377  125914.150080   \n",
       "15       7.742014e+07  101923113.0       577.0   77954.305026  111304.550321   \n",
       "16       6.790927e+07   90222851.0       823.0   47577.639125  118305.706859   \n",
       "17       5.993617e+07   83095122.0      1138.0   40418.720562   75646.899374   \n",
       "18       5.927409e+07   80247514.0       265.0  109459.656604  148031.437933   \n",
       "19       4.886059e+07   58573575.0      1122.0   24486.719251   33383.796074   \n",
       "20       4.934577e+07   64313132.0       536.0   54251.886194  125214.452212   \n",
       "21       3.612821e+07   46605208.0        70.0   95460.300000  177068.179267   \n",
       "22       3.953276e+07   50801309.0       361.0   40188.700831   70611.924513   \n",
       "19       4.812576e+07   58003349.0       240.0   25084.708333   24626.483113   \n",
       "21       4.389644e+07   46665124.0       147.0   53972.210884   90592.560662   \n",
       "22       4.493636e+07   50445090.0        91.0   66167.461538   89383.026598   \n",
       "7        1.076262e+08  156640654.0        71.0  102202.507042  181652.537774   \n",
       "8        1.204696e+08  145066685.0       356.0   96684.362360  150044.306593   \n",
       "X        1.298868e+08  156025710.0       811.0   65726.053021  145339.541124   \n",
       "Y        2.300546e+07   25988069.0        61.0   54031.918033  117153.357329   \n",
       "\n",
       "                                                          \n",
       "            min       25%      50%        75%        max  \n",
       "seqname                                                   \n",
       "1         191.0   9072.00  24448.0   58905.00  1551956.0  \n",
       "2         313.0  14108.50  43435.0  101049.25  1463630.0  \n",
       "3         224.0  14727.50  41582.5   93488.50  1743269.0  \n",
       "4         134.0  16185.75  48647.0  112178.75  1506191.0  \n",
       "5         236.0  14679.00  43041.0  102746.50  1553082.0  \n",
       "6         155.0   9442.00  25999.0   79145.50  1987246.0  \n",
       "7         257.0  12206.00  32601.0   86587.50  2304996.0  \n",
       "8         662.0   9758.25  29936.0   77911.25  2059619.0  \n",
       "9         341.0   7955.00  27983.0   84100.75  2298756.0  \n",
       "10        239.0  14789.00  38975.0   90568.00  1851180.0  \n",
       "11        209.0   6022.00  16346.0   47649.00  2172910.0  \n",
       "12         77.0  10998.75  29043.5   71513.00  1258479.0  \n",
       "13        398.0  19597.50  46371.0  112931.50  1475061.0  \n",
       "14        112.0   8988.75  25254.0   77939.25  1697918.0  \n",
       "15        272.0  15130.00  38526.0   97223.00   969298.0  \n",
       "16        272.0   7688.00  19768.0   51933.50  2473538.0  \n",
       "17         62.0   7108.00  17021.5   43190.75  1161877.0  \n",
       "18        215.0  26960.00  52441.0  137560.00  1195706.0  \n",
       "19        741.0   7882.25  15428.0   29851.25   485248.0  \n",
       "20        305.0   9654.25  23600.5   57606.00  2057828.0  \n",
       "21        369.0  15188.75  38115.5   87553.00  1216866.0  \n",
       "22        476.0   9252.00  20297.0   41999.00   760615.0  \n",
       "19        838.0   8399.25  16572.5   33944.25   144080.0  \n",
       "21        252.0   8047.00  27325.0   65406.00   836159.0  \n",
       "22       1379.0  11442.00  32987.0   83496.50   550561.0  \n",
       "7         650.0  18552.00  39586.0  113986.00  1195210.0  \n",
       "8         347.0  13074.75  44002.5  116458.75  1214011.0  \n",
       "X         347.0   5930.00  19277.0   66260.00  2241932.0  \n",
       "Y         242.0   2814.00  14471.0   42272.00   741998.0  \n",
       "\n",
       "[29 rows x 24 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_per_chr.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d726e9a-8d40-444b-8929-11af5c4dac9d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
